packages/automerge-repo/src/DocumentPhasor.ts

```diff
@@ -0,0 +1,829 @@
+import debug from "debug"
+import { DocHandle, SyncInfo } from "./DocHandle.js"
+import { headsAreSame } from "./helpers/headsAreSame.js"
+import {
+  DocMessage,
+  DocumentUnavailableMessage,
+  EphemeralMessage,
+  MessageContents,
+  RequestMessage,
+  SyncMessage,
+} from "./network/messages.js"
+import { DocumentId, PeerId } from "./types.js"
+import { next as Automerge } from "@automerge/automerge"
+import { decode } from "./helpers/cbor.js"
+import { Request } from "./phase/Request.js"
+import { Loading } from "./phase/Loading.js"
+import { Ready } from "./phase/Ready.js"
+import { Unavailable } from "./phase/Unavailable.js"
+
+interface SpawnPhasorArgs<T> {
+  ourPeerId: PeerId
+  documentId: DocumentId
+  initialState: Automerge.Doc<T> | undefined
+  networkReady: boolean
+}
+
+// A `DocumentPhasor` manages the lifecycle of a document as it transitions
+// through several different phases. The general intent here is that the
+// `DocumentPhasor` is a pretty self contained state machine, it does not
+// share state with other parts of the system.
+//
+// To use a DocumentPhasor you create the phasor, then you either:
+//
+// a) Feed it `DocEvent`s via `DocumentPhasor.handleEvent` informing the phasor of changes to
+//   the environment (such as a new connection)
+// b) Modify the document managed by the DocumentPhasor using `DocumentPhasor.localChange`
+//
+// Both methods will return a `DocEffects`, which is a description of the things which have
+// changed as a result of the operation.
+//
+// ## Phases
+//
+// The phasor manages the following phases:
+//
+// * loading
+// * requesting
+// * ready
+// * unavailable
+//
+// ## Caller Responsibilities
+//
+// ### Loading the document
+//
+// The `DocEffects` returned by the `handleEvent` method will indicate whether
+// it is necessary to load the document from storage. When an event returns
+// `DocEffects.beginLoad: true` then the caller should load the document. When
+// the load completes the caller should provide the result of the load via the
+// "load_complete" event.
+//
+// ### Network Ready
+//
+// Initially a phasor is in a "network not ready" state. This means that it will not mark
+// documents as unavailable until the "network_ready" event has been received. Callers
+// should dispatch this event as soon as the network subsystem says it is ready.
```

Comment:
I don't think this refactor should change the state machine, but in the long term, it would be nice to support the network going in and out of the "ready" state. Have you considered how that fits in here?